packages/shared/src/hooks/useBookmarkPost.ts

```diff
@@ -1,82 +1,218 @@
-import { useContext } from 'react';
-import { useMutation } from 'react-query';
-import request from 'graphql-request';
+import { useCallback, useContext } from 'react';
+import { useMutation, useQueryClient } from 'react-query';
 import { graphqlUrl } from '../lib/config';
 import {
   ADD_BOOKMARKS_MUTATION,
+  Post,
   REMOVE_BOOKMARK_MUTATION,
+  ReadHistoryPost,
 } from '../graphql/posts';
 import AnalyticsContext from '../contexts/AnalyticsContext';
-import { MutateFunc } from '../lib/query';
 import { useToastNotification } from './useToastNotification';
-import { AnalyticsEvent } from './analytics/useAnalyticsQueue';
+import { useRequestProtocol } from './useRequestProtocol';
+import AuthContext from '../contexts/AuthContext';
+import { updatePostCache } from './usePostById';
+import { AuthTriggers } from '../lib/auth';
+import { AnalyticsEvent, Origin } from '../lib/analytics';
+import {
+  PostAnalyticsEventFnOptions,
+  optimisticPostUpdateInFeed,
+  postAnalyticsEvent,
+} from '../lib/feed';
+import { FeedItem, PostItem, UpdateFeedPost } from './useFeed';
 
-type UseBookmarkPostParams<T> = {
-  onBookmarkMutate: MutateFunc<T>;
-  onRemoveBookmarkMutate: MutateFunc<T>;
-  onBookmarkTrackObject?: () => AnalyticsEvent;
-  onRemoveBookmarkTrackObject?: () => AnalyticsEvent;
+export type ToggleBookmarkProps = {
+  origin: Origin;
+  post: Post | ReadHistoryPost;
+  opts?: PostAnalyticsEventFnOptions;
 };
-type UseBookmarkPostRet<T> = {
-  bookmark: (variables: T) => Promise<void>;
-  bookmarkToast: (targetState: boolean) => void;
-  removeBookmark: (variables: T) => Promise<void>;
+
+export const bookmarkMutationKey = ['post', 'mutation'];
+
+export type UseBookmarkPostMutationProps = {
+  id?: string;
+  mutation?: string;
+  payload?: Record<string, unknown>;
 };
 
-export default function useBookmarkPost<
-  T extends { id: string } = { id: string },
->({
-  onBookmarkMutate,
-  onRemoveBookmarkMutate,
-  onBookmarkTrackObject,
-  onRemoveBookmarkTrackObject,
-}: UseBookmarkPostParams<T>): UseBookmarkPostRet<T> {
-  const { trackEvent } = useContext(AnalyticsContext);
+export type UseBookmarkPostRollback = () => void;
+
+export type BookmarkProps = {
+  id: string;
+};
+
+export type UseBookmarkPostProps = {
+  onMutate?: (
+    props: UseBookmarkPostMutationProps,
+  ) => Promise<UseBookmarkPostRollback> | UseBookmarkPostRollback | undefined;
+  variables?: unknown;
```

Comment:
Don't we know what types we expect here?

Reply:
You're right. Going to adjust it a bit and throw in the type `MutationKey`

Reply:
Refactored it to have a type now